---
title: "Distance"
author: "Serena"
date: "10 marzo 2016"
output: html_document
---


```{r}
source("https://bioconductor.org/biocLite.R")
biocLite()
library(devtools)
biocLite("genomicsclass/tissuesGeneExpression")
library(tissuesGeneExpression)
data(tissuesGeneExpression)
dim(e) ##gene expression data
table(tissue) ##tissue[i] corresponds to e[,i]
```

#1
How many biological replicates for hippocampus?
```{r}
length(which (tissue=="hippocampus"))
```
#2
What is the distance between samples 3 and 45?
```{r}
dist(t(e[,c(3,45)]))
```
#3
What is the distance between gene 210486_at and 200805_at
```{r}
dist(e[c("210486_at","200805_at"),])
```
#4
If I run the command (don't run it!):

d = as.matrix( dist(e) )

how many cells (number of rows times number of columns) will this matrix have?
```{r}
nrow(e)*nrow(e)
```
#5
Compute the distance between all pair of samples:
```{r}
d = dist( t(e) )
```
Read the help file for dist.

How many distances are stored in d? Hint: What is the length of d?
```{r}
length(d)
```
#6
Why is the answer to exercise 5 not ncol(e)^2?

A) R made a mistake there.
B) Distances of 0 are left out.
#C) Because we take advantage of symmetry: only lower triangular matrix is stored thus only ncol(e)*(ncol(e)-1)/2 values.
D) Because it is equalnrow(e)^2
