  ---
title: "Lung Adenocarcinoma"
author: "Francesco Asnicar"
date: "10 March 2016"
output: 
    html_document: 
      fig_caption: yes
      number_sections: yes
      toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=FALSE)
```


```{r}
suppressMessages(library(Biobase))

# setwd("~/University/Courses/Applied Statistic for High-Throughput Biology/AppStatTrento/projects/franz89_Lung_Adenocarcinoma")

eset <- readRDS("LUAD/LUAD_RNAseq2.rds") ## eset and filename are examples
# exprs(eset) ## provides the expression data
# pData(eset) ## provides "phenotype" data, in this case clinical data
# sampleNames(eset) ## provides patient identifiers
# featureNames(eset) ## provides gene symbols

subtypes <- read.csv("LUAD/LUAD_subtypes.csv", as.is=TRUE)
sub.eset <- exprs(eset)[, substr(sampleNames(eset), 1, 12) %in% tolower(subtypes$sampleId)]
pheno <- pData(eset)[substr(row.names(pData(eset)), 1, 12) %in% tolower(subtypes$sampleId), ]
```


# First, introduce yourself to the dataset

## Basic countings
How many genes and samples are there?

Genes: `r dim(sub.eset)[1]`

Samples: `r dim(sub.eset)[2]`


## At least one count
For each gene, calculate the number of samples where it has at least one count.

```{r}
genes.count <- apply(sub.eset, 1, function(x) sum(x>0))
sum(genes.count > 0)
```

How many genes have at least one count in at least 10% of the samples?

```{r}
sum(genes.count > dim(sub.eset)[2]/10)
```

For each gene, calculate the % of samples in which it has 1 or more counts.  Plot these percentages as a histogram.
<!-- Present a histogram of the % of samples in which each gene has 1 or more counts. -->

```{r, echo=FALSE}
# genes.perc <- genes.count / ncol(sub.eset) * 100
# hist(genes.perc, main="", xlab="% samples", ylab="gene counts")
hist(genes.count/dim(sub.eset)[2]*100, main="", xlab="% samples", ylab="gene counts")
```


## Boxplot
Create a single boxplot showing the expression counts of 16 randomly selected genes.

```{r}
set.seed(17)
samples.16 <- sub.eset[sample(dim(sub.eset)[1], size=16),]
boxplot(as.vector(samples.16))
```

Do the data appear normally distributed?

**ANSWER:** No, data appear distributed like a Poisson. In fact, we have a long tails on the right side, while most of te values appear to be near the 0, where we can see the median is located.


## Boxplot normalized
Create another boxplot for the same genes, after adding 1 and taking base-2 logarithm of the expression counts.

```{r}
samples.16.log2 <- log2(samples.16 + 1)
boxplot(as.vector(samples.16.log2), ylim=c(0, 14))
```

Is this more normally distributed? If so, use this transformed version of the dataset for further exploratory analysis.

**ANSWER:** Yes, with this transformation almost all the distributions of the expression values appear to be more normally distributed.


## Histograms
For another look at these gene expression distributions, make 16 histograms on a 4x4 grid.

```{r, echo=FALSE}
library(rafalib)
rafalib::mypar(4, 4)

for (i in 1:16) {
  hist(samples.16.log2[,i], main=row.names(samples.16.log2)[i], xlab="", ylim=c(0, 6))
}
```


## More plots
Using whichever transformation between 4 and 5 resulted in a more nearly normal distribution, create a boxplot showing expression values for each sample.

```{r, echo=FALSE}
rafalib::mypar(1, 1, las=2)
boxplot(t(samples.16.log2), ylim=c(0, 14))
```

Do any samples look like outliers?

**ANSWER**: Yes, clearly the VWC2L gene is an outlier, even after the log transofrmation.


## Show any other exploratory analysis you try and find useful.

```{r, echo=FALSE}
rafalib::mypar(1, 2, las=1)
qqnorm(as.vector(samples.16), main="16 samples")
qqline(as.vector(samples.16))

qqnorm(as.vector(samples.16.log2), main="16 samples log2")
qqline(as.vector(samples.16.log2))
```

```{r, echo=FALSE}
rafalib::mypar(1, 1, las=2)
boxplot(scale(t(samples.16.log2), center=TRUE), ylim=c(-6, 7), main="Centered boxplots")
```

# A visual look for cluster structure

## Screeplot
Remove any genes that have non-zero counts in fewer than 10% of the samples. Using the more normally distributed transformation of the data, perform a Principal Components Analysis (PCA). A standard base R solution for PCA is the `prcomp()` function. Be sure to include variance scaling for the genes, which can be done manually or as an option to the `prcomp()` function. Show the screeplot of your PCA, and comment.

```{r, echo=FALSE}
sub.eset.10perc <- log2(sub.eset[(genes.count > ncol(sub.eset)/10),] + 1)
sub.eset.10perc.std <- t(scale(t(sub.eset.10perc), scale=FALSE))
p <- prcomp(sub.eset.10perc.std)
rafalib::mypar()
plot(p$sdev^2 / sum(p$sdev^2)*100, ylab="% variance explained", main="Screeplot")
screeplot(p)

```

**ANSWER**:  `r sum((p$sdev^2 / sum(p$sdev^2)*100)[1:2])`% of the variance is explained by the first two principal components.


## PCA plots
Make two side-by-side plots of PC2 vs PC1. In the first, do not distinguish between subtypes. In the second, use color and data point shapes to distinguish between the reported subtypes. Do you see any obvious structure or clustering of the samples? Be VERY sure you have correctly aligned patient barcodes in the ExpressionSet and the barcodes file, for example using identical().

```{r}
rafalib::mypar(1, 2, las=1)
plot(p$rotation[, 1:2], main="PC1 vs. PC2", xlab="PC1 8.14%", ylab="PC2 6.7%")
colors <- palette()[1:length(unique(subtypes$expression_subtype))]
plot(p$rotation[, 1:2], main="PC1 vs. PC2 by subtypes", xlab="PC1 8.14%", ylab="PC2 6.7%", col=colors, pch=as.integer(factor(colors)))
legend("bottomright", legend=unique(subtypes$expression_subtype), col=colors, lwd=1, pch=1:length(unique(subtypes$expression_subtype)), lty=NA)
```


## Extra analysis

### MDS

```{r}
rafalib::mypar()
d.cor <- as.dist(1-cor(sub.eset.10perc.std))
mds <- cmdscale(d.cor)
plot(mds, main="", col=colors)
legend("topleft", legend=unique(subtypes$expression_subtype), col=colors, lwd=1, pch=1, lty=NA)
```


# Differential Expression

## ANOVA
Use the DESeq2 Bioconductor package to assess differential expression of each gene across the subtypes, using a one-way Analysis of Variance (ANOVA). Create a histogram of raw p-values.

```{r, echo=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
suppressMessages(library(DESeq2))

control <- factor(subtypes$expression_subtype)

dds <- DESeqDataSetFromMatrix(matrix(as.integer(sub.eset), nrow=20501, ncol=230), DataFrame(control), ~ control)
#dds <- dds[rowSums(counts(dds)) > 1, ] # filtering

dds1 <- DESeq(dds)
dds.res <- results(dds1)

mypar(1, 2)
hist(dds.res$pvalue, main="Raw p-values", ylim=c(0, 15000), xlab="p-values")
```

Create a second histogram of F-statistics, adding a vertical red line at the critical F-statistic corresponding to a Bonferroni-corrected p-value of 0.05.

```{r, echo=FALSE}
hist(dds.res$padj, main="Adjusted p-values", ylim=c(0, 15000), xlab="p-values")
abline(v=0.05, col="red")
```

Repeat these two histograms, after permuting the subtype labels. Comment on the shapes of the histograms. Note that the DESeq2 differential expression should be performed on untransformed expression values.

```{r, echo=FALSE}
perm.control <- factor( base::sample(subtypes$expression_subtype) )

perm.dds <- DESeqDataSetFromMatrix(matrix(as.integer(sub.eset), nrow=20501, ncol=230), DataFrame(perm.control), ~ perm.control)
#perm.dds <- perm.dds[rowSums(counts(perm.dds)) > 1, ] # filtering

perm.dds1 <- DESeq(perm.dds)
perm.dds.res <- results(perm.dds1)

mypar(1, 2)
hist(perm.dds.res$pvalue, main="Premuted raw p-values", ylim=c(0, 15000), xlab="p-values")
hist(perm.dds.res$padj, main="Permuted adjusted p-values", ylim=c(0, 15000), xlab="p-values")
abline(v=0.05, col="red")
```


## Heatmap
Create a clustered heatmap of the top 1000 differentially expressed genes, ranked by p-value, with rows (genes) scaled to unit variance (row scaling is an automatic option available in most heatmap plotting functions. Create a column sidebar, for example by using the annotation_col argument of `pheatmap::pheatmap()`.

```{r}
library(pheatmap)
suppressMessages(library(genefilter))
library(RColorBrewer)

idx <- order(dds.res$padj)[1:1000]

annot <- data.frame(subtype=control)
rownames(annot) <- colnames(sub.eset)
annot.col <- list(subtype=RColorBrewer::brewer.pal(n=length(levels(control)), "Set1"))
names(annot.col[[1]]) <- unique(subtypes$expression_subtype) # labels for the legend

colors <- RColorBrewer::brewer.pal(8, "Set2")
pheatmap::pheatmap(log2(sub.eset[idx, ]+1), scale="row", show_rownames=F, show_colnames=F, col=colors, main=NA, annotation_col=annot, annotation_colors=annot.col)
```


# Unsupervised Clustering

Attempt to reproduce the unsupervised clustering methods reported by the paper for your cancer type. Create a "confusion matrix" showing the concordance between subtype calls produced by your method, and as provided by TCGA. Report on the methods you used, and how similar your results are to those reported.


## Paper's unsupervised clustering

Previously validated gene expression subtypes of lung adenocarcinoma [11,12] were detected in the TCGA lung adenocarcinoma cohort. Gene expression data were gene median centered. Using previously published predictor centroids [12,13] subtype was assigned to each TCGA tumor specimen using a nearest centroid predictor [13], limiting to the genes common to the predictor and the TCGA cohort (n = 489) and using Pearson correlation as the similarity metric, with the maximum correlation coefficient providing the subtype prediction for a tumor (subtype calls in [14]). To empirically assess the quality of these subtype detections similar to earlier studies [1,12], expression of the predictor genes was compared between the TCGA cohort and the previously published Wilkerson et al. cohort (Supplementary Figure 7B). Subtype expression patterns were highly concordant between the cohorts, indicating that the subtypes are a similar stratification of the TCGA cohort as in earlier cohorts. 

```{r}
eset.median <- apply(sub.eset, 1, function(x){
  x - median(x)
})
eset.median.scaled <- t(scale(eset.median, center=FALSE))

centroids <- read.csv("wilkerson.2012.LAD.predictor.centroids.csv", as.is=T)
centroids <- centroids[centroids$X %in% rownames(eset.median.scaled),]

predicted.subtypes <- apply(eset.median.scaled[centroids$X,], 2, function(s){
  lbls <- c("TRU", "Proximal Proliferative", "Proximal Inflammatory")
  cors <- c(cor(s, centroids$bronchioid), # TRU
            cor(s, centroids$magnoid), # Proximal Proliferative
            cor(s, centroids$squamoid)) # Proximal Inflammatory
  idx <- order(cors, decreasing=T)[1]
  c( cors[idx], lbls[idx] )
})

control <- factor(predicted.subtypes[2,])
annot <- data.frame(subtype=control)
rownames(annot) <- colnames(eset.median.scaled)
annot.col <- list(subtype=c("black", "green", "red"))
names(annot.col[[1]]) <- unique(predicted.subtypes[2,]) # labels for the legend

mypar()
colors <- RColorBrewer::brewer.pal(11, "RdBu")
pheatmap::pheatmap(eset.median.scaled[centroids$X,], scale="row", show_rownames=F, show_colnames=F, col=colors, annotation_col=annot, annotation_colors=annot.col)
```

I tried to follow what reported in the Supplementary Material of the paper.
In particular, I managed to scale the data by the median of the genes.
I downloaded the centroids, as referenced in [13], and assign the subtype to the samples according to the Pearson correlation coefficient computed between the sample profile and the three centroids.
However, the result I obtained is substantially different from the one presented in the Sumpplementary Material.
For start, the mRNA expression leves they present are limited to the (-1, 1) interval, while the values that I have range from almost (-15, 15).
This, for sure, affects the displayed colors in the heatmap.
The other difference is in the subtypes coloring.
While they present a very neat and discrete sepration between samples, the results I got suggest that such clustering is not present.
It could be that this is due to the fact that they removed such values below -1 or above 1, affecting in this way the samples clustering.
A strange thing of the Supplemtary Figure 7b, is the absence of the clutering tree, especially for the samples.


## Hierarchical clustering

```{r}
mypar()
hc.cor <- hclust( dist( 1-cor(sub.eset) ) )
rafalib::myplclust(hc.cor, labels=control, lab.col=as.numeric(factor( control )), cex=0.5, main=NA)
legend("top", legend=levels(factor(control)), lty=1,  lw=3, col=1:length(unique(control)), bty="n")
abline(h=5, col="red")
```


## K-means

Computed with respect to the `batch_number` field in the phenotype data.

```{r}
clusters.cor <- cutree(hc.cor, k=5)
set.seed(1)
km <- kmeans( dist( t(sub.eset.10perc) ), centers=length(unique(pheno$batch_number)) )
table(Kmeans=km$cluster, Hclust=clusters.cor)
chisq.test(km$cluster, pheno$batch_number, simulate.p.value=TRUE)
```

Computed with respect to the `expression_subtype` type.

```{r}
set.seed(1)
km <- kmeans( dist( t(sub.eset.10perc) ), centers=length(unique(subtypes$expression_subtype)) )
table(Kmeans=km$cluster, Hclust=clusters.cor)
chisq.test(km$cluster, pheno$batch_number, simulate.p.value=TRUE)
```
