---
title: "Papillary Renal Cell Carcinoma"
author: "Calogero Zarbo"
date: "March 4, 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
eset <- readRDS("KIRP_RNAseq2.rds")   ## eset and filename are examples
library(Biobase)
expr <- data.frame(exprs(eset))  ##provides the expression data
pheno <- pData(eset)  ##provides “phenotype” data, in this case clinical data
samples <- sampleNames(eset) ##provides patient identifiers
features <- featureNames(eset) ##provides gene symbols
```

First, introduce yourself to the dataset:

1) How many genes and samples are there?

2) For each gene, calculate the number of samples where it has at least one count. How many genes have at least one count in at least 10% of the samples? Present a histogram of the % of samples in which each gene has 1 or more counts.
```{r}
gene_presence <- data.frame(Gene = c(),N_Samples=c(), Percentage = c())
for (i in 1:length(features)){
  c_sample = 0
  gene_name = features[i] 
  for (j in 1:length(samples)){
    gene_count = expr[i,j]
    # print(class(gene_count))
    if (gene_count > 0){
        c_sample = c_sample + 1
    }
  }
  b <- data.frame(Gene = gene_name, N_Samples = c_sample, Percentage = c_sample/length(samples))
  gene_presence <- rbind(gene_presence,b)
  
}

hist(gene_presence$Percentage)

```


3) Create a single boxplot showing the expression counts of 16 randomly selected genes. Do the data appear normally distributed?
```{r}
set.seed(1)
smpl <- sample(expr, 16)
smpl <- data.frame(t(smpl))
single_col = data.frame(Expr = c())

for (i in 1:ncol(smpl)){
  col <- data.frame(Expr = smpl[,i])
  single_col <- rbind(single_col, col)
  
}

boxplot(single_col)
```



4) Create another boxplot for the same genes, after adding 1 and taking base-2 logarithm of the expression counts. Is this more normally distributed? If so, use this transformed version of the dataset for further exploratory analysis.
```{r}

log_single_col <- log(1+single_col)
boxplot(log_single_col)

```


5) For another look at these gene expression distributions, make 16 histograms on a 4x4 grid.
```{r}
old.par <- par(mfrow=c(4, 4))
for (i in 1:ncol(smpl)){
  col <- smpl[,i]
  hist(col, xlab=colnames(smpl)[i])
}
```


6) using whichever transformation between 4 and 5 resulted in a more nearly normal distribution, create a boxplot showing expression values for each sample. Do any samples look like outliers?
```{r}
old.par <- par(mfrow=c(4,4))

 for (i in 1:nrow(smpl)){
   row <- smpl[i,]
   boxplot(row, xlab=rownames(smpl)[i])
 }

print ("Yes some samples look like outliers")
```



Show any other exploratory analysis you try and find useful.

```{r}
old.par <- par(mfrow=c(4,4))
normality <- data.frame(Gene = c(), value = c(), p_value=c())
for(i in 1:length(features)){
  feat <- data.frame(t(expr[i,]))
  name <- features[i]
  if(sum(feat) > 0){
    lshap <- lapply(feat, shapiro.test)
    b <- data.frame(Gene = name,value = lshap[[1]]$statistic, p_value = lshap[[1]]$p.value)
    normality <- rbind(normality, b)
  }

}

hist(normality$value)
```

```{r}
normality_log <- data.frame(Gene = c(), value = c(), p_value=c())
for(i in 1:length(features)){
  feat <- data.frame(t(expr[i,]))
  name <- features[i]
  if(sum(feat) > 0){
    lshap <- lapply(log(1+feat), shapiro.test)
    b <- data.frame(Gene = name,value = lshap[[1]]$statistic, p_value = lshap[[1]]$p.value)
    normality_log <- rbind(normality_log, b)
  }

}
hist(normality_log$value)

```

A visual look for cluster structure

7) Remove any genes that have non-zero counts in fewer than 10% of the samples. Using the more normally distributed transformation of the data, perform a Principal Components Analysis (PCA). A standard base R solution for PCA is the prcomp() function. Be sure to include variance scaling for the genes, which can be done manually or as an option to the prcomp() function. Show the screeplot of your PCA, and comment.

```{r}
old.par <- par(mfrow=c(1,1))
gene_to_remove <- data.frame(Name = c())
for (i in 1:length(features)){
  c_sample = 0
  gene_name = features[i] 
  for (j in 1:length(samples)){
    gene_count = expr[i,j]
    # print(class(gene_count))
    if (gene_count == 0){
        c_sample = c_sample + 1
    }
  }
  if (c_sample /length(samples) > 0.1 ){
    b <- data.frame(Name = gene_name)
    gene_to_remove <- rbind(gene_to_remove,b)
  }
  
}
```

```{r}
expr_clean <- expr[ , !(names(expr) %in% gene_to_remove)]
log_expr_clean <-  log(1+expr_clean)

e.standardize.fast <- t(apply(log_expr_clean, 1, function(x) x - mean(x) ))
scale <- princomp(e.standardize.fast, cor=TRUE)
no_scale <- princomp(e.standardize.fast)

plot(no_scale$sdev^2 / sum(scale$sdev^2)*100, ylab="% variance explained", main="Scale screen plot")

```



8) Make two side-by-side plots of PC2 vs PC1. In the first, do not distinguish between subtypes. In the second, use color and data point shapes to distinguish between the reported subtypes. Do you see any obvious structure or clustering of the samples? Be VERY sure you have correctly aligned patient barcodes in the ExpressionSet and the barcodes file, for example using identical().

```{r}


  plot(scale$loadings[, 1:2], xlab="PC1", ylab="PC2", 
       main=paste("plot of scale$loadings[, 1:2]",colnames(pheno)[1],sep=" "),
       col=factor(pheno[,1]), pch=as.integer(factor(pheno[,1])))
  legend("topleft", legend=levels(factor(pheno[,1])), col=1:length(unique(pheno[,1])),
         pch=1:length(unique(pheno[,1])), bty='n')
  
  print("There is no clear structure in the dataset")
  
```




Differential Expression

9) Use the DESeq2 Bioconductor package to assess differential expression of each gene across the subtypes, using a one-way Analysis of Variance (ANOVA). Create a histogram of raw p-values. Create a second histogram of F-statistics, adding a vertical red line at the critical F-statistic corresponding to a Bonferroni-corrected p-value of 0.05. Repeat these two histograms, after permuting the subtype labels. Comment on the shapes of the histograms. Note that the DESeq2 differential expression should be performed on untransformed expression values.
```{r}

library(DESeq2)
library(SummarizedExperiment)


subtypes <- read.csv2("KIRP_subtypes.csv", sep=" ")

subtypes_names <- c()

for (i in 1:length(subtypes$bcr_patient_barcode.mRNA)) {
  s_name <- subtypes$bcr_patient_barcode.mRNA[i]
  s_name <- gsub('-', '.', s_name)
  s_name <- substr(s_name, 1, regexpr(",",s_name)-1)
  s_name <- paste(s_name, "01", sep=".")
  s_name <- tolower(s_name)

  subtypes_names <- rbind(subtypes_names,s_name)
  
}

expr_clean_subtypes <- expr_clean[,(colnames(expr_clean) %in% subtypes_names)]

rounded_expr_clean<- round(expr_clean_subtypes, digits = 0) #It's wrong but I didn't found the way to get the counts

condition <- factor(subtypes$clusters)
```

```{r}


dseq_ds <- DESeqDataSetFromMatrix(rounded_expr_clean, DataFrame(condition), ~ condition)

diff_exprs_genes <- DESeq(dseq_ds, test = c("Wald", "LRT"), fitType = c("parametric", "local",
"mean"))

res <- results(diff_exprs_genes)
resOrdered <- res[order(res$padj),]

hist(resOrdered$padj)
stat_borderline = resOrdered$stat[length(which (resOrdered$padj <= 0.05))]
hist(resOrdered$stat)
abline(v=stat_borderline,col="red")
```

```{r}
#Permuting Labels
condition_2 <- factor(data.matrix(subtypes$clusters[sample(length(subtypes$clusters))]))
dseq_ds_2<- DESeqDataSetFromMatrix(rounded_expr_clean, DataFrame(condition_2), ~ condition_2)

diff_exprs_genes_2 <- DESeq(dseq_ds_2, test = c("Wald", "LRT"), fitType = c("parametric", "local",
"mean"))

res_2 <- results(diff_exprs_genes_2)
resOrdered_2 <- res_2[order(res_2$padj),]

hist(resOrdered_2$padj)
stat_borderline_2 = resOrdered_2$stat[length(which (resOrdered_2$padj <= 0.05))]
hist(resOrdered_2$stat)
abline(v=stat_borderline_2,col="red")

print("By permutating the labels way less genes will be discriminative for that phenotype")
```


10) Create a clustered heatmap of the top 1000 differentially expressed genes, ranked by p-value, with rows (genes) scaled to unit variance (row scaling is an automatic option available in most heatmap plotting functions. Create a column sidebar, for example by using the annotation_col argument of pheatmap::pheatmap().

```{r}
library(pheatmap)
library(RColorBrewer)

df_resordered <- data.frame(resOrdered)
first_genes <- rownames(df_resordered)[1:1000]
first_expr_clean_subtypes <- expr_clean_subtypes[(rownames(expr_clean_subtypes) %in% first_genes),]

palette <- RColorBrewer::brewer.pal(12, "Paired")

annot <- data.frame(type = condition)
annot.col <- list(type = RColorBrewer::brewer.pal(n=3, "PiYG"))


hc_euc <- dist(first_expr_clean_subtypes)
hc <- (hclust(hc_euc))

pheatmap::pheatmap(t(first_expr_clean_subtypes), scale = "row", show_rownames = F, show_colnames = F, main = "Hierarchical clustering", color = palette, annotation_col = annot, annotation_colors = annot.col, cluster_cols = hc)


```


Unsupervised Clustering

11) Attempt to reproduce the unsupervised clustering methods reported by the paper for your cancer type. Create a “confusion matrix” showing the concordance between subtype calls produced by your method, and as provided by TCGA. Report on the methods you used, and how similar your results are to those reported.

"The shared probe set between HM27 and HM450 platforms (N=25,978) were used for this
analysis. We removed probes that contained any masked data due to detection p value, repeats
and SNPs and non-uniquely mapped probes (n=23,381 remaining)."
They didn't provide any parameter for the selection.

"The R package FDb.InfiniumMethylation.hg19 was used to obtain chromosomal locations of the probes and 961
X-linked probes were removed from the analysis. A standard deviation for each probe was
calculated across the 161 tumors in the final data freeze. We chose probes that were
unmethylated in the 50 normal samples (mean beta value <0.1) that had a standard deviation of
greater than 0.15 (n=343) in the tumors for the clustering."
Not being familiar with the library, would be nice to have the exact line of codes they used to perform the selection here. Not present in the supplementary materials.

"Hierarchical clustering with Ward’s
method was used to cluster the 161 data freeze samples, and the clustering dendrogram was cut
at k=3 to yield three clusters after assessing cluster strength and stability. One of the clusters
(cluster number = 3) exhibited extensive hypermethylation across thousands of CpG loci and was
renamed CpG island methylator phenotype (CIMP)."



```{r}

hc_pearson <- as.dist(1-(cor(expr_clean_subtypes)))
hc <- (hclust(hc_pearson, method="ward.D2"))
rafalib::myplclust(hc, labels=condition, lab.col=as.numeric(factor(condition)), main="1 - PCC Distance Cluster Dendrogram", cex=0.5)
legend("topright", legend=levels(factor(condition)), pch = 19, col=1:length(unique(condition)), bty="n")
abline(h=1, lw=2, lty=1)

hclusters <- cutree(hc, h=1)
table(true=condition, cluster=hclusters)


pheatmap::pheatmap(t(expr_clean_subtypes), scale = "row", show_rownames = F, show_colnames = F, main = "Hierarchical clustering", color = palette, annotation_col = annot, annotation_colors = annot.col, cluster_cols = hc, cluster_rows = hclusters, cutree_cols = 3)
```

```{r}
hc_pearson <- as.dist(1-(cor(expr_clean_subtypes)))
hc <- (hclust(hc_pearson, method="ward.D"))
rafalib::myplclust(hc, labels=condition, lab.col=as.numeric(factor(condition)), main="1 - PCC Distance Cluster Dendrogram", cex=0.5)
legend("topright", legend=levels(factor(condition)), pch = 19, col=1:length(unique(condition)), bty="n")
abline(h=1, lw=2, lty=1)

hclusters <- cutree(hc, h=1)
table(true=condition, cluster=hclusters)


pheatmap::pheatmap(t(expr_clean_subtypes), scale = "row", show_rownames = F, show_colnames = F, main = "Hierarchical clustering", color = palette, annotation_col = annot, annotation_colors = annot.col, cluster_cols = hc, cluster_rows = hclusters, cutree_cols = 3)
```

Trying to build heatmap using the first genes who passed the differential expression test using corrected p-value.

```{r}
first_genes_good <- rownames(df_resordered)[1:length(which (resOrdered$padj <= 0.05))]
first_expr_clean_subtypes_good <- expr_clean_subtypes[(rownames(expr_clean_subtypes) %in% first_genes_good),]

hc_pearson <- as.dist(1-(cor(first_expr_clean_subtypes_good)))
hc <- (hclust(hc_pearson, method="ward.D2"))
rafalib::myplclust(hc, labels=condition, lab.col=as.numeric(factor(condition)), main="1 - PCC Distance Cluster Dendrogram", cex=0.5)
legend("topright", legend=levels(factor(condition)), pch = 19, col=1:length(unique(condition)), bty="n")
abline(h=1, lw=2, lty=1)

hclusters <- cutree(hc, h=1)
table(true=condition, cluster=hclusters)


pheatmap::pheatmap(t(first_expr_clean_subtypes_good), scale = "row", show_rownames = F, show_colnames = F, main = "Hierarchical clustering", color = palette, annotation_col = annot, annotation_colors = annot.col, cluster_cols = hc, cluster_rows = hclusters, cutree_cols = 3)

```

I don't see a strong similarity between the two results, but to replicate the same unsupervised clustering the proper R line codes used for the preprocessing are necessary.

